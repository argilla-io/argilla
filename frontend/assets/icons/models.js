/*
 * coding=utf-8
 * Copyright 2021-present, the Recognai S.L. team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable */
var icon = require('vue-svgicon')
icon.register({
  'models': {
    width: 18,
    height: 17,
    viewBox: '0 0 18 17',
    data: '<g fill-rule="nonzero" _fill="none"><path pid="0" d="M3.5 16.5a3 3 0 110-6 3 3 0 010 6zm0-1a2 2 0 100-4 2 2 0 000 4zM14.5 6.5a3 3 0 110-6 3 3 0 010 6zm0-1a2 2 0 100-4 2 2 0 000 4zM3.5 6.5a3 3 0 110-6 3 3 0 010 6zm0-1a2 2 0 100-4 2 2 0 000 4z" _fill="#F38959"/><path pid="1" _fill="#F48E5F" d="M4 11H3V6h1zM5.5 4.5v-1H12v1z"/></g>'
  }
})
