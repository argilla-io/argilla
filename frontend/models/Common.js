/*
 * coding=utf-8
 * Copyright 2021-present, the Recognai S.L. team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class BaseRecord {
  id;
  metadata;
  prediction;
  annotation;
  predicted;
  status;

  selected;

  constructor({
    id,
    metadata,
    prediction,
    annotation,
    predicted,
    status,
    selected,
  }) {
    this.id = id;
    this.metadata = metadata;
    this.prediction = prediction;
    this.annotation = annotation;
    this.predicted = predicted;
    this.status = status;
    this.selected = selected || false;
  }

  recordTitle() {
    throw new Error("Method 'recordTitle()' must be implemented.");
  }
}

class BaseSearchQuery {
  multi_label;
  predicted_as;
  annotated_as;
  annotated_by;
  predicted_by;
  status;
  predicted;
  metadata;
  text;
  from;
  limit;

  constructor({
    predicted_as,
    annotated_as,
    annotated_by,
    predicted_by,
    status,
    predicted,
    metadata,
    text,
  }) {
    this.predicted_as = predicted_as;
    this.annotated_as = annotated_as;
    this.annotated_by = annotated_by;
    this.predicted_by = predicted_by;
    this.status = status;
    this.predicted = predicted;
    this.metadata = metadata;
    this.text = text;
  }
}

class BaseSearchResults {
  total;
  records;
  aggregations;

  constructor({ total, records, aggregations }) {
    this.total = total || 0;
    this.records = records;
    this.aggregations = aggregations || {};
  }
}

export { BaseRecord, BaseSearchQuery, BaseSearchResults };
