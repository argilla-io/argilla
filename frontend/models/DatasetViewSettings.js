/*
 * coding=utf-8
 * Copyright 2021-present, the Recognai S.L. team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Model } from "@vuex-orm/core";

class Pagination extends Model {
  static entity = "pagination_settings";

  static fields() {
    return {
      id: this.string(null),
      size: this.number(5),
      page: this.number(1),
      pageSizeOptions: this.attr([1, 5, 10, 20]),
    };
  }

  get from() {
    return (this.page - 1) * this.size;
  }
}

export default class DatasetViewSettings extends Model {
  static entity = "view_settings";

  static fields() {
    return {
      id: this.string(null),
      pagination: this.hasOne(Pagination, "id"),
      annotationEnabled: this.boolean(false),
      loading: this.boolean(false),
    };
  }
}

export { Pagination, DatasetViewSettings };
