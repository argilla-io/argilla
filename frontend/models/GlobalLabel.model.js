import { Model } from "@vuex-orm/core";

class GlobalLabel extends Model {
  static entity = "globalLabels";
  static primaryKey = ["text", "color_id"];

  static fields() {
    return {
      id: this.uid(),
      order: this.number(null),
      color_id: this.attr(null),
      text: this.string(null),
      dataset_id: this.attr(null),
      is_activate: this.attr(false),
      shortcut: this.attr(null),
    };
  }

  static mutators() {
    return {
      text(value) {
        return value.toUpperCase();
      },
    };
  }
}

const formatDatasetIdForGlobalLabelModel = (dataset_id) => dataset_id.join(".");

export { GlobalLabel, formatDatasetIdForGlobalLabelModel };
