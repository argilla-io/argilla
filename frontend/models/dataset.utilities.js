import { getTokenClassificationDatasetById } from "@/models/tokenClassification.queries";
import { getTextClassificationDatasetById } from "@/models/textClassification.queries";
import { getText2TextDatasetById } from "@/models/text2text.queries";

const getDatasetFromORM = (datasetId, datasetTask) => {
  try {
    return getTaskDatasetById(datasetId, datasetTask);
  } catch (err) {
    console.error(err);
    return null;
  }
};

const getTaskDatasetById = (datasetId, datasetTask) => {
  let datasetById = null;
  switch (datasetTask.toUpperCase()) {
    case "TEXTCLASSIFICATION":
      datasetById = getTextClassificationDatasetById(datasetId);
      break;
    case "TOKENCLASSIFICATION":
      datasetById = getTokenClassificationDatasetById(datasetId);
      break;
    case "TEXT2TEXT":
      datasetById = getText2TextDatasetById(datasetId);
      break;
    default:
      throw new Error(`ERROR Unknown task: ${datasetTask}`);
  }
  return datasetById;
};

export { getDatasetFromORM };
