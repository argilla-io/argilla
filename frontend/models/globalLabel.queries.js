import {
  GlobalLabel as GlobalLabelModel,
  formatDatasetIdForGlobalLabelModel,
} from "./GlobalLabel.model";

const upsertLabelsInGlobalLabelModel = (labels) => {
  GlobalLabelModel.insertOrUpdate({
    data: labels,
  });
};
const deleteAllGlobalLabelModel = () => {
  GlobalLabelModel.deleteAll();
};

const getAllLabelsByDatasetId = (datasetId, sortBy = "order", asc = true) => {
  const directionSort = asc ? "asc" : "desc";
  const joinedDatasetId = formatDatasetIdForGlobalLabelModel(datasetId);
  const labels = GlobalLabelModel.query()
    .where("dataset_id", joinedDatasetId)
    .orderBy(sortBy, directionSort)
    .get();

  return labels;
};

const getAllLabelsTextByDatasetId = (
  datasetId,
  sortBy = "order",
  asc = true
) => {
  return getAllLabelsByDatasetId(datasetId, sortBy, asc).reduce(
    (acc, curr) => acc.concat(curr.text),
    []
  );
};

const countLabelsByDatasetId = (datasetId) => {
  const joinedDatasetId = formatDatasetIdForGlobalLabelModel(datasetId);
  const numberOfLabels = GlobalLabelModel.query()
    .where("dataset_id", joinedDatasetId)
    .count();

  return numberOfLabels;
};

const insertNewGlobalLabel = ({ datasetId, newLabel, isActivate = false }) => {
  const joinedDatasetId = formatDatasetIdForGlobalLabelModel(datasetId);
  const numberOfLabels = countLabelsByDatasetId(datasetId);
  GlobalLabelModel.insert({
    data: {
      id: newLabel,
      text: newLabel,
      dataset_id: joinedDatasetId,
      order: numberOfLabels,
      color_id: numberOfLabels,
      shortcut: numberOfLabels < 9 ? String(numberOfLabels + 1) : null,
      is_activate: isActivate,
    },
  });
};

const isLabelTextExistInGlobalLabel = (datasetId, labelText) => {
  const joinedDatasetId = formatDatasetIdForGlobalLabelModel(datasetId);

  const compareWithCaseInsensitive = (value) =>
    value.toUpperCase() === labelText.toUpperCase();

  return GlobalLabelModel.query()
    .where("dataset_id", joinedDatasetId)
    .where("text", compareWithCaseInsensitive)
    .exists();
};

export {
  getAllLabelsByDatasetId,
  upsertLabelsInGlobalLabelModel,
  deleteAllGlobalLabelModel,
  insertNewGlobalLabel,
  getAllLabelsTextByDatasetId,
  isLabelTextExistInGlobalLabel,
};
