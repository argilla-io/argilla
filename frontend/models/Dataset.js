/*
 * coding=utf-8
 * Copyright 2021-present, the Recognai S.L. team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Model } from "@vuex-orm/core";
import DatasetViewSettings from "./DatasetViewSettings";

const USER_DATA_METADATA_KEY = "rubrix.recogn.ai/ui/custom/userData.v1";

class ObservationDataset extends Model {
  static entity = "datasets";

  static primaryKey = "name";

  static fields() {
    return {
      name: this.string(null),
      metadata: this.attr(null),
      tags: this.attr(null),
      task: this.string(null),
      created_at: this.string(null),
      last_updated: this.string(null),
      viewSettings: this.hasOne(DatasetViewSettings, "id"),
    };
  }

  get visibleRecords() {
    return this.results.records.slice(
      0,
      this.viewSettings.pagination.size
    );
  }
}

export { ObservationDataset, USER_DATA_METADATA_KEY };
