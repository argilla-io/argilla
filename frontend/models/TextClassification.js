/*
 * coding=utf-8
 * Copyright 2021-present, the Recognai S.L. team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ObservationDataset, USER_DATA_METADATA_KEY } from "./Dataset";
import { BaseRecord, BaseSearchQuery, BaseSearchResults } from "./Common";

class TextClassificationRecord extends BaseRecord {
  inputs;

  constructor({ inputs, explanation, multi_label, ...superData }) {
    super(superData);
    this.inputs = inputs;
    this.explanation = explanation;
    this.multi_label = multi_label;
  }

  recordTitle() {
    return this.inputs;
  }
}

class TextClassificationSearchQuery extends BaseSearchQuery {
  score;

  constructor(data) {
    const { score, confidence, ...superData } = data;
    super(superData);
    this.score = score;
    // TODO: remove backward compatibility
    if (confidence) {
      this.score = confidence;
    }
  }
}

class TextClassificationSearchResults extends BaseSearchResults {
  constructor({ total, records, aggregations }) {
    super({
      total,
      aggregations,
      records: (records || []).map(
        (record) => new TextClassificationRecord(record)
      ),
    });
  }
}

class TextClassificationDataset extends ObservationDataset {
  static entity = "text_classification";

  static fields() {
    return {
      ...super.fields(),
      query: this.attr({}, (data) => {
        return new TextClassificationSearchQuery(data);
      }),
      sort: this.attr([]),
      results: this.attr(
        {},
        (data) => new TextClassificationSearchResults(data)
      ),
      globalResults: this.attr({}),
    };
  }

  get labels() {
    const { labels } = (this.metadata || {})[USER_DATA_METADATA_KEY] || {};
    const aggregations = this.globalResults.aggregations;

    const uniqueLabels = [
      ...new Set(
        (labels || [])
          .filter((l) => l && l.trim())
          .concat(Object.keys(aggregations.annotated_as))
          .concat(Object.keys(aggregations.predicted_as))
      ),
    ];
    uniqueLabels.sort();
    return uniqueLabels;
  }
}

export {
  TextClassificationDataset,
  TextClassificationRecord,
  TextClassificationSearchResults,
  TextClassificationSearchQuery,
};
