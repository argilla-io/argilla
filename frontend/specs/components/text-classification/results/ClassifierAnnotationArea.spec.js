import { mount } from "@vue/test-utils";
import Component from "@/components/text-classifier/results/ClassifierAnnotationArea";

import { TextClassificationRecord } from "@/models/TextClassification";

const $route = {
  query: {},
};

function mountSidebar() {
  return mount(Component, {
    propsData: {
      record: new TextClassificationRecord({
        inputs: { text: "My text", multi_label: false },
      }),
      dataset: {
        task: "TextClassification",
        viewSettings: {
          annotationEnabled: false,
        },
        labels: ["A", "B"],
      },
    },
    mocks: {
      $route,
    },
  });
}

describe("ClassifierAnnotationArea", () => {
  let spy = jest.spyOn(console, "error");
  afterEach(() => spy.mockReset());

  test("Required property", () => {
    expect(() => {
      mount(Component);
      expect(spy).toBeCalledWith(
        expect.stringContaining('[Vue warn]: Missing required prop: "dataset"')
      );
    }).toThrowError(TypeError);
  });

  test("renders properly", () => {
    const wrapper = mountSidebar();
    expect(wrapper).toMatchSnapshot();
  });
});
