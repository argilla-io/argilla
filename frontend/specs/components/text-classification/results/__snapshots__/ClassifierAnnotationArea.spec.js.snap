// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassifierAnnotationArea renders properly 1`] = `
<div class="annotation-area">
  <!---->
  <div class="feedback-interactions">
    <classifierannotationbutton id="A" label="[object Object]" data-title="A" value="A" class="label-button"></classifierannotationbutton>
    <classifierannotationbutton id="B" label="[object Object]" data-title="B" value="B" class="label-button"></classifierannotationbutton>
    <!---->
  </div>
</div>
`;
